Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp

```diff
@@ -0,0 +1,712 @@
+                    !(zOrigin < configData.collisionRegionMin) &&
+                    !(zOrigin > configData.collisionRegionMax)) {
+                  // We keep counting duplets with atomic variables
+                  const auto ind = deviceCountBotDuplets[mid].fetch_add(1);
+                  deviceTmpIndBot[mid * B + ind] = bot;
+                }
+              }
+            });
+      });
+
+      q->submit([&](cl::sycl::handler& h) {
+        h.parallel_for<duplet_search_top_kernel>(
+            cl::sycl::range<2>{M, B}, [=](cl::sycl::id<2> idx) {
+              const auto mid = idx[0];
+              const auto top = idx[1];
+              // We check whether this thread actually makes sense (within
+              // bounds).
+              if (mid < M && top < T) {
+                const auto midSP = deviceMiddleSPs[mid];
+                const auto topSP = deviceTopSPs[top];
+
+                const auto deltaR = topSP.r - midSP.r;
+                const auto cotTheta = (topSP.z - midSP.z) / deltaR;
+                const auto zOrigin = midSP.z - midSP.r * cotTheta;
+
+                if (!(deltaR < configData.deltaRMin) &&
+                    !(deltaR > configData.deltaRMax) &&
+                    !(cl::sycl::abs(cotTheta) > configData.cotThetaMax) &&
+                    !(zOrigin < configData.collisionRegionMin) &&
+                    !(zOrigin > configData.collisionRegionMax)) {
+                  // We keep counting duplets with atomic access.
+                  const auto ind = deviceCountTopDuplets[mid].fetch_add(1);
+                  deviceTmpIndTop[mid * T + ind] = top;
+                }
+              }
+            });
+      });
+    }  // sync
+
+    q->memcpy(deviceNumTopDuplets, deviceCountTopDuplets, M * sizeof(uint32_t));
+
+    //*********************************************//
+    // *********** DUPLET SEARCH - END *********** //
+    //*********************************************//
+
+    // retrieve results from counting duplets
+    {
+      std::vector<uint32_t> countBotDuplets(M);
+      std::vector<uint32_t> countTopDuplets(M);
+
+      q->memcpy(countBotDuplets.data(), deviceCountBotDuplets,
+                M * sizeof(std::atomic_uint32_t));
+      q->memcpy(countTopDuplets.data(), deviceCountTopDuplets,
+                M * sizeof(std::atomic_uint32_t));
+      q->wait();
+
+      // use std::parital_sum instead?
+      for (auto i = 1; i < M + 1; ++i) {
+        sumBotCompUptoMid[i] +=
+            sumBotCompUptoMid[i - 1] + countBotDuplets[i - 1];
+        sumTopCompUptoMid[i] +=
+            sumTopCompUptoMid[i - 1] + countTopDuplets[i - 1];
+        sumBotTopCombined[i] += sumBotTopCombined[i - 1] +
+                                countTopDuplets[i - 1] * countBotDuplets[i - 1];
+      }
+
+      edgesBottom = sumBotCompUptoMid[M];
+      edgesTop = sumTopCompUptoMid[M];
+      edgesComb = sumBotTopCombined[M];
+
+      if (edgesBottom == 0 || edgesTop == 0)
+        return;
```

Comment:
What happens with all the allocated device memory when we return here? Yepp, it leaks... :frowning:

Reply:
Omg, thanks.